### PR TITLE
Refactor DB Package to Enable Multiple Blocks/States at Slots

### PR DESCRIPTION
**Pull request by rauljordan**
This PR eliminates all instances of `BlockBySlot`, as forks can cause several blocks to be saved at the same slot in our runtime. Our code does not account for this and tries to fetch whatever block exists at a certain slot, possibly leading to deviations and discrepancies, and in worst case scenarios, intense reorgs.

This PR refactors the `db` package to allow for the above, and also fixes the block children algorithms throughout the repo.

### CODE REVIEW

File: beacon-chain/db/block.go [hunk 1]
```diff
@@ -243,57 +241,85 @@ func (db *BeaconDB) UpdateChainHead(ctx context.Context, block *pb.BeaconBlock,
...
 	})
 }
 
-// BlockBySlot accepts a slot number and returns the corresponding block in the main chain.
-// Returns nil if a block was not recorded for the given slot.
-func (db *BeaconDB) BlockBySlot(ctx context.Context, slot uint64) (*pb.BeaconBlock, error) {
-	_, span := trace.StartSpan(ctx, "BeaconDB.BlockBySlot")
+// CanonicalBlockBySlot accepts a slot number and returns the corresponding canonical block.
+func (db *BeaconDB) CanonicalBlockBySlot(ctx context.Context, slot uint64) (*pb.BeaconBlock, error) {
+	_, span := trace.StartSpan(ctx, "BeaconDB.BlocksBySlot")
```
Comment (prestonvanloon) [inline beacon-chain/db/block.go]:
```suggestion
	_, span := trace.StartSpan(ctx, "BeaconDB.CanonicalBlockBySlot")
```

File: beacon-chain/db/block.go [hunk 2]
```diff
@@ -243,57 +241,85 @@ func (db *BeaconDB) UpdateChainHead(ctx context.Context, block *pb.BeaconBlock,
...
 
-		enc := blockBkt.Get(blockRoot)
-		if enc == nil {
-			return nil
-		}
+	return block, err
+}
+
+// BlocksBySlot accepts a slot number and returns the corresponding blocks in the db.
+// Returns nil if no blocks were recorded for the given slot.
```
Comment (prestonvanloon) [inline beacon-chain/db/block.go]:
```suggestion
// Returns empty list if no blocks were recorded for the given slot.
```

File: beacon-chain/db/state.go [hunk 3]
```diff
@@ -44,7 +47,7 @@ func (db *BeaconDB) InitializeState(ctx context.Context, genesisTime uint64, dep
 	blockRoot, _ := hashutil.HashBeaconBlock(genesisBlock)
 	// #nosec G104
 	blockEnc, _ := proto.Marshal(genesisBlock)
-	zeroBinary := encodeSlotNumber(0)
+	zeroBinary := encodeSlotNumberRoot(params.BeaconConfig().GenesisSlot, blockRoot)
```
Comment (shayzluf) [inline beacon-chain/db/state.go]: why zero contains encode slot number root of (params.BeaconConfig().GenesisSlot, blockRoot)?

Comment (rauljordan) [reply beacon-chain/db/state.go]: Good q @shayzluf , we can have multiple blocks saved at a single slot, so this way, we can prefix search for them in the DB. I will add more documentation to explain this.

File: beacon-chain/db/state.go [hunk 4]
```diff
@@ -271,11 +285,10 @@ func (db *BeaconDB) FinalizedState() (*pb.BeaconState, error) {
 
 // HistoricalStateFromSlot retrieves the state that is closest to the input slot,
 // while being smaller than or equal to the input slot.
-func (db *BeaconDB) HistoricalStateFromSlot(ctx context.Context, slot uint64) (*pb.BeaconState, error) {
+func (db *BeaconDB) HistoricalStateFromSlot(ctx context.Context, slot uint64, blockRoot [32]byte) (*pb.BeaconState, error) {
```
Comment (nisdas) [inline beacon-chain/db/state.go]: maybe rename this to `HistoricalStateFromBlock` ?

**Pull request merged**